_posts/2018-07-20-exploring-state-management-in-react-container-components.md

```diff
@@ -0,0 +1,125 @@
+  };
+
+  handleAddItemClick = () => {
+    this.props.addItem(this.state.newItem);
+
+    this.setState({ newItem: "" });
+  };
+
+  render() {
+    return (
+      <ul>
+        <h2>Cart</h2>
+        <input
+          type="text"
+          onChange={({target:{value: newItem}}) => 
+            this.setState({newItem})}
+          value={this.state.newItem}
+        />
+        <button onClick={this.handleAddItemClick}>
+          Add item
+        </button>
+
+        {this.props.items.map((item, idx) => (
+          <li key={idx}>
+            <button onClick={this.props.removeItem(item)}>
+              Remove
+            </button>
+            {item}, &ensp;
+            {this.props.discounts[idx]}% off!
+          </li>
+        ))}
+      </ul>
+    );
+  }
+}
+
+class CartContainer extends React.Component {
+  state = {
+    items: [],
+    discounts: []
+  };
+
+  componentDidUpdate(prevProps, prevState) {
+    if (prevState.items.length !== this.state.items.length) {
+      const itemsQuery = this.state.items.join(",");
+
+      ajax(`https://discountdb.com/?items=${itemsQuery}`)
+        .then(discounts => this.setState({ discounts }));
+    }
+  }
+
+  addItem = item => {
+    this.setState(prevState => ({
+      items: [...prevState.items, item]
+    }));
+  };
+
+  removeItem = item => () => {
+    this.setState(prevState => ({
+      items: prevState.items.filter(
+        (i, idx, prevItems) => idx !== prevItems.indexOf(item)
+      )
+    }));
+  };
+
+  render() {
+    return (
+      <div>
+        <Cart
+          items={this.state.items}
+          discounts={this.state.discounts}
+          addItem={this.addItem}
+          removeItem={this.removeItem}
+        />
+      </div>
+    );
+  }
+}
+```
+&nbsp;
+
+## Benefits
+
+Partitioning your components on the state boundary layer will help reduce complexity by simply having less code to work with at a time, while still staying "inside React".
+
+I really think this pattern starts to pay dividends when you have a lifecycle method, like `componentDidUpdate`, doing a lot of asynchronous work (like making HTTP requests). Given the asynchronous nature, this sort of code tends to be very difficult to test (with both automated unit test and manual testing), so breaking this stateful logic into separate components helps keeps you sane and your code focused. 
+
+It's helpful to remind yourself that when unit testing a React component, you are essentially testing the `render` function. Given the inputs (`props`), what is the output? You've probably noticed tests are painful to write if there is a lot of setup, especially if the setup is required for a feature of the component that you aren't even testing.
+
+Keeping your components small and focused will go a long way for keeping yourself happy and productive!
+
+Play with the code by following the link below!
+
+[![Edit wo7y9voowk](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/wo7y9voowk)
+
```

Comment:
The end is kind of abrupt, especially given L21 above where you promise `a few different patterns`. To be clear I'm remarking on the inconsistency, not the length.

Reply:
Good catch, this was originally going to be a comprehensive post, covering several patterns, but the first part ended up taking more time/being longer than I expected, so I decided I would instead turn it into a series.

Do you have any ideas on how to conclude with that in mind? I was having trouble figuring out how much to say in the "in the next post" part.

>Next time, we'll work through...

I also wasn't sure whether I should title the post `Exploring State Management in React: Part I: Container Components`. Sort of a mouthful

Reply:
What about "Using Container Components for React State Management" You can still make it a series, but gives you flexibility of where to go next. Maybe move the Redux discussion to the end as a teaser. Or keep it where it is as joking/sarcasm. "So you think you can do React without Redux?? Think again!!" kinda vibe.

Reply:
I'd be careful not to promise too much - there's no telling what you'll write for your next article.